test/framework/control_plane.go

```diff
@@ -34,84 +34,191 @@ import (
+// Can be updated by consumer to change default values.
+var (
+	// DefaultTimeout is used for the default timeout for every `Eventually` call.
+	DefaultTimeout = 10 * time.Minute
+
+	// DefaultPollInterval is used as the default for how frequently `Eventually` polls.
+	DefaultPollInterval = 10 * time.Second
 )
 
-// ControlplaneClusterInput defines the necessary dependencies to run a multi-node control plane cluster.
-type ControlplaneClusterInput struct {
-	Management        ManagementCluster
-	Cluster           *clusterv1.Cluster
-	InfraCluster      runtime.Object
-	Nodes             []Node
-	MachineDeployment MachineDeployment
-	RelatedResources  []runtime.Object
-	CreateTimeout     time.Duration
-	DeleteTimeout     time.Duration
+// Interfaces to scope down client.Client
 
-	ControlPlane    *controlplanev1.KubeadmControlPlane
-	MachineTemplate runtime.Object
+// Getter can get resources.
+type Getter interface {
+	Get(ctx context.Context, key client.ObjectKey, obj runtime.Object) error
+}
+
+// Creator can creates resources.
+type Creator interface {
+	Create(ctx context.Context, obj runtime.Object, opts ...client.CreateOption) error
+}
+
+// Lister can lists resources.
+type Lister interface {
+	List(ctx context.Context, list runtime.Object, opts ...client.ListOption) error
+}
+
+// Deleter can delete resources.
+type Deleter interface {
+	Delete(ctx context.Context, obj runtime.Object, opts ...client.DeleteOption) error
+}
+
+// ListDeleter combines interfaces.
+type ListerDeleter interface {
+	Lister
+	Deleter
+}
+
+// Durations allow the customization of a test timeout and polling interval.
+// They can be customized for each test and/or the default can be changed by updating the package globals.
+type Durations struct {
+	Timeout      time.Duration
+	PollInterval time.Duration
 }
 
-// SetDefaults defaults the struct fields if necessary.
-func (input *ControlplaneClusterInput) SetDefaults() {
-	if input.CreateTimeout == 0 {
-		input.CreateTimeout = 10 * time.Minute
+// SetDefaults is a convenience for library consumers so they never have to specify values for this type.
+func (d *Durations) SetDefaults() {
+	if d.Timeout == 0 {
+		d.Timeout = DefaultTimeout
 	}
+	if d.PollInterval == 0 {
+		d.PollInterval = DefaultPollInterval
+	}
+}
 
-	if input.DeleteTimeout == 0 {
-		input.DeleteTimeout = 5 * time.Minute
+// CreateRelatedResourcesInput is the input type for CreateRelatedResources.
+type CreateRelatedResourcesInput struct {
+	Creator          Creator
+	RelatedResources []runtime.Object
+	Durations
+}
+
+// CreateRelatedResources is used to create runtime.Objects.
+func CreateRelatedResources(ctx context.Context, input CreateRelatedResourcesInput) {
+	input.SetDefaults()
+
+	By("creating related resources")
+	for i := range input.RelatedResources {
+		obj := input.RelatedResources[i]
+		By(fmt.Sprintf("creating a/an %s resource", obj.GetObjectKind().GroupVersionKind()))
+		Eventually(func() error {
```

Comment:
I'm also wondering if it would make sense to add an offset to either the input struct or the function arguments to allow us to plumb that through to `EventuallyWithOffset` and make errors returned from these helpers a bit cleaner: https://onsi.github.io/gomega/#making-assertions-in-helper-functions

We could even automatically increment the offset if we have helper methods that invoke additional helper methods.

Reply:
I like the idea, that's an additional feature we didn't support before but I'd be way interested in adding it. Could you file an issue for this so we can track it?

Reply:
Done. https://github.com/kubernetes-sigs/cluster-api/issues/2305